File: src/Arduino/keyboardControl.py
import serial
import keyboard
import time
import threading


def send_command(command):
    arduino.write(command.encode())


def read_serial():
    global servo_position
    while True:
        if arduino.in_waiting > 0:  # Check if there's data to read
            line = arduino.readline().decode('utf-8').rstrip()  # Read the line, decode to string, and strip newlines
            with position_lock:
                servo_position = line  # Update the shared variable
            print(f"\r{line}", end="")  # Print the received data
        time.sleep(0.1 / 50)  # Small delay to prevent CPU overuse


def ask_angle():
    while True:
        if keyboard.is_pressed('a'):
            send_command('a')  # Wait for key release with a short timeout
            time.sleep(.2)

def main():
    # print("Press the up/down arrow keys to control the servo. Press 'q' to quit.")
    wait = 1 / 60  # 60 Hz

    try:
        while True:
            if keyboard.is_pressed('up'):
                send_command('u')
                # print("Sending 'u' command to increase position")
                time.sleep(wait)  # Send command at intervals (adjust as needed)
            elif keyboard.is_pressed('down'):
                send_command('d')
                #                 print("Sending 'd' command to decrease position")
                time.sleep(wait)  # Send command at intervals (adjust as needed)
            elif keyboard.is_pressed('q'):
                send_command('q')
                print("\nExiting...")
                break
    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        arduino.close()


if __name__ == '__main__':
    # Set up the serial connection (change 'COM3' to your actual Arduino port)
    arduino = serial.Serial('COM4', 9600, timeout=1)

    # Shared variable and lock
    servo_position = None
    position_lock = threading.Lock()

    # Create and start the serial reading thread
    serial_thread = threading.Thread(target=read_serial)
    serial_thread.daemon = True  # This makes the thread exit when the main program exits
    serial_thread.start()

    angle_thread = threading.Thread(target=ask_angle)
    angle_thread.daemon = True  # This makes the thread exit when the main program exits
    angle_thread.start()

    try:
        # Run the main task
        main()
    except KeyboardInterrupt:
        print("Exiting...")
    finally:
        arduino.close()  # Close the serial connection when done

File: src/Arduino/keyboardControl.cpp
//
// Created by evert on 27/05/2024.
//

#include "keyboardControl.h"


int keyboard(Servo servo, int servoPos) {
    if (Serial.available() > 0) {
        char command = Serial.read(); // Read the incoming byte

        if (command == 'u') { // 'u' for up arrow key
            servoPos = 75; // Increase the servo position
        } else if (command == 'd') { // 'd' for down arrow key
            servoPos = 5; // Decrease the servo position
        } else if (command == 'q') { // 'q' for quit
            servo.write(5);
            Serial.println("Quitting the program");
            delay(1000); // Wait for 1 second
            exit(0); // Exit the program
        } else if (command == 'a') { // 'a' for angle
            Serial.print("Servo Position: ");
            Serial.println(servoPos);
        }

        servo.write(servoPos); // Move the servo to the new position
    }
    return servoPos;
}